components/feature/media/src/main/java/mozilla/components/feature/media/MediaSessionFeature.kt

```diff
@@ -0,0 +1,92 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.media
+
+import android.app.Activity
+import android.content.Context
+import android.content.pm.ActivityInfo
+import androidx.core.content.ContextCompat
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.map
+import mozilla.components.browser.state.state.SessionState
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.feature.media.service.AbstractMediaSessionService
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+
+/**
+ * Feature implementation that handles MediaSession state changes
+ *
+ * @param activity the application's [Activity].
+ * @param mediaServiceClass the media service class will handle the media playback state
+ * @param store Reference to the browser store where tab state is located.
+ * @param applicationContext the application's [Context].
+ */
+class MediaSessionFeature(
+    val activity: Activity,
+    val mediaServiceClass: Class<*>,
+    val store: BrowserStore,
+    val applicationContext: Context = activity.applicationContext
+
+) : LifecycleAwareFeature {
+    private var scope: CoroutineScope? = null
+
+    /**
+     * Starts the feature.
+     */
+    override fun start() {
+        scope = store.flowScoped { flow ->
+            flow.ifChanged { it.tabs + it.customTabs }.map { state ->
+                state.tabs.filter { it.mediaSessionState != null } +
+                    state.customTabs.filter { it.mediaSessionState != null }
+            }.collect { states ->
+                process(states)
+            }
```

Comment:
You are over-subscribing/processing here, I think:

* The first `ifChanged()` will fire for *any* change in a tab or custom tab, e.g. if the loading percentage of any tab changes then this chain continues. Primarily `ifChanged()` exists to limit flow processing to only the changes that are relevant for the use case.
* You map the values after the changes check. You may want to consider first mapping to the values you need and then checking if they changed.
* Mixing the two cases (fullscreen and starting the service) may make this more complicated than needed. Even if you want to use just one feature class, internally you should consider delegating to two internal classes that only handle one case.

Reply:
I see.  I'll investigate if I should split into two feature classes.

Reply:
Updated with split classes